gpkit/nomials/array.py

```diff
@@ -191,33 +178,27 @@ def sum(self, *args, **kwargs):
         "Returns a sum. O(N) if no arguments are given."
         if args or kwargs or all(l == 0 for l in self.shape):
             return np.ndarray.sum(self, *args, **kwargs)
-        cs = []
-        units = self.units
-        exps = []
+        hmap = NomialMap()
+        hmap.units = self.units
         it = np.nditer(self, flags=['multi_index', 'refs_ok'])
         while not it.finished:
             i = it.multi_index
             it.iternext()
             if isinstance(mag(self[i]), Numbers):
-                # somehow a number got in here
                 if mag(self[i]) == 0:
                     continue
-                else:
-                    cs.append(mag(self[i]))
-                    exps.append({})
-            cs.extend(mag(self[i].cs).tolist())
-            exps.extend(self[i].exps)
-        if units:
-            cs *= units
-        return Signomial(exps, cs)
+                else:  # number manually inserted by user
+                    hmap[EMPTY_EXP] = mag(self[i]) + hmap.get(EMPTY_EXP, 0)
```

Comment:
Yeah I think I tried doing that way back when and got confused about whether / how to inherit from `defaultdict(float)` without overriding getitem. Seems like a great refactor to do in the future, along with removing zeros on the fly!